tornado/test/websocket_test.py

```diff
@@ -831,14 +835,79 @@ def on_ping(self, data):
 
     @gen_test
     def test_client_ping(self):
-        ws = yield self.ws_connect("/", ping_interval=0.01)
+        ws = yield self.ws_connect("/", ping_interval=0.01, ping_timeout=0)
         for i in range(3):
             response = yield ws.read_message()
             self.assertEqual(response, "got ping")
         # TODO: test that the connection gets closed if ping responses stop.
         ws.close()
 
 
+class ServerPingTimeoutTest(WebSocketBaseTestCase):
+    def get_app(self):
+        handlers = []
+
+        class PingHandler(TestWebSocketHandler):
+            def initialize(self, close_future=None, compression_options=None):
+                # capture the handler instance so we can interrogate it later
+                nonlocal handlers
+                handlers.append(self)
+                return super().initialize(
+                    close_future=close_future,
+                    compression_options=compression_options,
+                )
+
+        app = Application([("/", PingHandler)])
+        app._handlers = handlers
+        return app
+
+    @staticmethod
+    def delay_pong(ws, delay):
+        """Delay the client's "pong" response by <delay> seconds."""
+        def wrapper(fcn):
+            def _inner(oppcode: int, data: bytes):
+                # print(f'$ {oppcode} {data}')  # uncomment to debug
+                if oppcode == 0xA:  # NOTE: 0x9=ping, 0xA=pong
+                    from time import sleep
+                    sleep(delay)
```

Comment:
Blocking `time.sleep` seems problematic here - it's not a realistic simulation of what happens in production (but now I see why you did the time checking below). 

How about this as a testing technique: instead of `delay_pong`, do `drop_ping` which replaces `protocol.write_ping` with a no-op (you can use `unittest.mock.patch` for this). Then pings don't get sent (even though the ping coroutine thinks they are), pongs don't get sent, and the connection should get closed on its own.

Reply:
I wanted to test the actual timings as closely as I could, but agree, it's a bit bunk.

I've changed this from delaying the client pong, to just suppressing the client pong (no need to suppress the server ping).

I haven't used `patch` for this as it's an instance method (so doesn't need to be reset post testing) and it's easier to target just the pongs with a wrapper.